test/cart_test_ep_credits.py

```diff
@@ -0,0 +1,127 @@
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# -*- coding: utf-8 -*-
+
+"""
+
+cart EP credits test
+
+Usage:
+
+Execute from the install/$arch/TESTING directory.
+
+python3 test_runner scripts/cart_test_ep_credits.yml
+
+To use valgrind memory checking
+set TR_USE_VALGRIND in cart_test_ep_credits.yml to memcheck
+
+To use valgrind call (callgrind) profiling
+set TR_USE_VALGRIND in cart_test_ep_credits.yml to callgrind
+
+"""
+
+import os
+from socket import gethostname
+import time
+import commontestsuite
+
+class TestEpCredits(commontestsuite.CommonTestSuite):
+    """ Execute EP credits tests """
+
+    def setUp(self):
+        """setup the test"""
+        self.get_test_info()
+        log_mask = os.getenv("D_LOG_MASK", "INFO")
+        log_file = self.get_cart_long_log_name()
+        crt_phy_addr = os.getenv("CRT_PHY_ADDR_STR", "ofi+sockets")
+        ofi_interface = os.getenv("OFI_INTERFACE", "eth0")
+        ofi_share_addr = os.getenv("CRT_CTX_SHARE_ADDR", "0")
+        ofi_ctx_num = os.getenv("CRT_CTX_NUM", "0")
+        self.pass_env = ' -x D_LOG_MASK={!s} -x D_LOG_FILE={!s}' \
+                        ' -x CRT_PHY_ADDR_STR={!s}' \
+                        ' -x OFI_INTERFACE={!s} -x CRT_CTX_SHARE_ADDR={!s}' \
+                        ' -x CRT_CTX_NUM={!s}'.format(
+                            log_mask, log_file, crt_phy_addr, ofi_interface,
+                            ofi_share_addr, ofi_ctx_num)
+
+    def tearDown(self):
+        """tear down the test"""
+        self.logger.info("tearDown begin")
+        self.logger.info("tearDown end\n")
+
+    def test_ep_credits(self):
+        """ep credits test"""
+        testmsg = self.shortDescription()
+
+        servers = self.get_server_list()
+        #run test even if server list is not provided.
+        if servers:
+            all_servers = ','.join(servers)
+            hosts = ''.join([' -H ', all_servers])
+        else:
+            hosts = ''.join([' -H ', gethostname().split('.')[0]])
+
+        srv_args = 'tests/test_ep_cred_server' + \
+            ' -n cred_group -s'
+
+        srv_proc = self.launch_bg(testmsg, '1', self.pass_env, hosts, srv_args)
+
+        time.sleep(2)
+
+        if srv_proc is None:
+            self.fail("Failed to launch cred server; return code %s" \
+                       % srv_proc.returncode)
+
+        # Important: 0 should be last in tuple, as client shuts down server
+        # for credit=0 case
+        credits_to_test = (1, 5, 10, 20, 255, 0)
+
+        for credit in credits_to_test:
+            cli_cmd = 'tests/test_ep_cred_client -a cred_group -c %d -b 20' \
```

Comment:
the -b number should always be more than the -c number in order to overflow the inflight queue.

Reply:
to overflow the queue yes, but i wanted to also test behaviors of other situations